### ISSUE TITLE
Please read & provide the following

### ISSUE THREAD

**Issue by jpmedinagl** [facebookresearch/detectron2#4290]
# Issue: Validation loss isn't decreasing or changing. I'm wondering if I implemented it correctly. 
The image below shows the total loss and validation loss versus iteration number.
The network is trained for 90,000 iterations and there are only two classes.

![image](https://user-images.githubusercontent.com/104694495/171215553-ee50f369-3466-47c1-895a-c62f172333f2.png)

I update my validation loss after every 5000 iterations because thats when the model is re-evaluated. I was also wondering if it would be better to make the validation loss update more frequently. I also don't know how to make the validation loss start as soon as the machine starts learning (or if its even possible)

1. Below is my trainer file and under that is the file that I run to train the network:
```
import os
from torch.utils.tensorboard import SummaryWriter
from detectron2.utils.events import EventWriter, get_event_storage
from detectron2.data import transforms as T
from detectron2.engine.hooks import HookBase
from detectron2.utils.logger import log_every_n_seconds
from detectron2.data import DatasetMapper, build_detection_test_loader, build_detection_train_loader
import detectron2.utils.comm as comm
import logging
import torch
import time
import datetime
import numpy as np
from detectron2.engine import DefaultTrainer
from detectron2.evaluation import COCOEvaluator

def build_train_aug(cfg):
    
    augs = [T.RandomBrightness(0.5, 2),
            T.RandomContrast(0.5, 2),
            T.RandomRotation([90, 180, 270, 360], expand=True, sample_style='choice'),
            T.RandomFlip(prob=0.5, horizontal=False, vertical=True)
            # T.ResizeScale(min_scale=0.2, max_scale=5, target_height=1, target_width=1)
            # T.RandomSaturation ?
            ]
    # T.RandomApply
    return augs

class Trainer(DefaultTrainer):
    @classmethod
    def build_evaluator(cls, cfg, dataset_name, output_folder=None):
        if output_folder is None:
            output_folder = os.path.join(cfg.OUTPUT_DIR, "inference")
        return COCOEvaluator(dataset_name, cfg, False, output_folder)

    def build_hooks(self):
        hooks = super().build_hooks()
        hooks.insert(-1,LossEvalHook(
            self.cfg.TEST.EVAL_PERIOD,
            self.model,
            build_detection_test_loader(
                self.cfg,
                self.cfg.DATASETS.TEST[0],
                DatasetMapper(self.cfg,True, augmentation=build_train_aug(self.cfg))
            )
        ))
        return hooks
    

class LossEvalHook(HookBase):
    def __init__(self, eval_period, model, data_loader):
        self._model = model
        self._period = eval_period
        self._data_loader = data_loader
    
    def _do_loss_eval(self):
        # Copying inference_on_dataset from evaluator.py
        total = len(self._data_loader)
        num_warmup = min(5, total - 1)
            
        start_time = time.perf_counter()
        total_compute_time = 0
        losses = []
        for idx, inputs in enumerate(self._data_loader):            
            if idx == num_warmup:
                start_time = time.perf_counter()
                total_compute_time = 0
            start_compute_time = time.perf_counter()
            if torch.cuda.is_available():
                torch.cuda.synchronize()
            total_compute_time += time.perf_counter() - start_compute_time
            iters_after_start = idx + 1 - num_warmup * int(idx >= num_warmup)
            seconds_per_img = total_compute_time / iters_after_start
            if idx >= num_warmup * 2 or seconds_per_img > 5:
                total_seconds_per_img = (time.perf_counter() - start_time) / iters_after_start
                eta = datetime.timedelta(seconds=int(total_seconds_per_img * (total - idx - 1)))
                log_every_n_seconds(
                    logging.INFO,
                    "Loss on Validation  done {}/{}. {:.6f} s / img. ETA={}".format(
                        idx + 1, total, seconds_per_img, str(eta)
                    ),
                    n=5,
                )
            loss_batch = self._get_loss(inputs)
            losses.append(loss_batch)
        mean_loss = np.mean(losses)
        self.trainer.storage.put_scalar('validation_loss', mean_loss)
        comm.synchronize()

        return losses
            
    def _get_loss(self, data):
        # How loss is calculated on train_loop 
        metrics_dict = self._model(data)
        metrics_dict = {
            k: v.detach().cpu().item() if isinstance(v, torch.Tensor) else float(v)
            for k, v in metrics_dict.items()
        }
        total_losses_reduced = sum(loss for loss in metrics_dict.values())
        return total_losses_reduced
        
        
    def after_step(self):
        next_iter = self.trainer.iter + 1
        is_final = next_iter == self.trainer.max_iter
        if is_final or (self._period > 0 and next_iter % self._period == 0):
            self._do_loss_eval()
        self.trainer.storage.put_scalars(timetest=12)

```
```
import os, sys
import matplotlib.pyplot as plt
import pandas as pd
from glob import glob
import numpy as np
import requests
import cv2
import json
import random
import torch

import detectron2

from detectron2 import model_zoo
from detectron2.engine import DefaultPredictor
from detectron2.config import get_cfg
from detectron2.utils.visualizer import Visualizer
from detectron2.data import MetadataCatalog
from detectron2.structures import BoxMode
from detectron2.engine import DefaultTrainer

from trainer_aug import Trainer

from detectron2.data import DatasetCatalog, MetadataCatalog
DatasetCatalog.clear()

# Then here I register my dataset

cfg = get_cfg()

# Right now using a model with no weights
cfg.merge_from_file(model_zoo.get_config_file('Misc/scratch_mask_rcnn_R_50_FPN_9x_gn.yaml'))
# "COCO-InstanceSegmentation/mask_rcnn_R_101_FPN_3x.yaml"))

cfg.DATASETS.TRAIN = ("data_train",)
cfg.DATASETS.TEST = ("data_small_val", )
cfg.TEST.EVAL_PERIOD = 5000 # Do evaluation after 5000 iteration
cfg.DATALOADER.NUM_WORKERS = 2

cfg.SOLVER.IMS_PER_BATCH = 2
cfg.SOLVER.BASE_LR = 0.001  # pick a good LR
cfg.SOLVER.MAX_ITER = 90000
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 2048   #(default: 512)
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 2  # Has two class

cfg.DATALOADER.ASPECT_RATIO_GROUPING = False

#For testing only
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.4

os.makedirs(model_loc + dir_name, exist_ok=True)
cfg.OUTPUT_DIR = model_loc + dir_name # + './output'

trainer = Trainer(cfg) ;
trainer.resume_or_load(resume=True)
trainer.train()

```
<put logs here>

## Expected behavior:

The expected behaviour is for the validation loss to decrease.

I'm not asking to make the model converge or work better. I just want to know why the validation loss isn't increasing and if I implemented the functions correctly.

## Environment:

I'm running on a super-computer that clears logs and everything everytime I run so I can't get the full environment, but I am using
torch 1.10
detectron2 0.6

**Comment by andreaceruti** [comment]
Hi @jpmedinagl I am facing the same issue, Logically it should means that you are overfitting from the very start of the training.
Also to me  is happening this problem where validation loss seems useless when selecting the best model. I am using this implementation found here on github issues, with some slight modification, but for me the best model is not found when using this implementation.
Instead what I do is to calculate mAP with the COCOEvaluator and then saving the model that reaches the best AP50 with the BestCheckpointer hook. 
So to sum up when I want to achieve the best COCO metrics on my MaskRCNN network I have to use the models returned by the BestCheckpointer or also the final trained network (after the whole iterations), instead the validation loss is misleading.
This could also be the reason why devs have not inserted it yet in the official hooks.

**Comment by jpmedinagl** [comment]
Hey @andreaceruti 
I ended up figuring out my problem, it had to do with the validation set that I was using. I had made 2 validation sets by accident, one was 'val' and the other was 'small_val'. Small_val had only one singular image. Thus as the training happened,
the validation was not decreasing because it was continually done on this singular image. I ended up having to change the validation set and after changing this the validation did decrease!

The code I changed from:
`cfg.DATASETS.TEST = ("data_small_val", )`
`cfg.DATASETS.TEST = ("data_val", )`

I see what you mean. I think I might try that. Thank you!
